crates/libp2p-networking/src/network/behaviours/dht/cache.rs

```diff
@@ -0,0 +1,322 @@
+
+    /// number of inserts since the last save
+    disk_parity_delta: Arc<AtomicU32>,
+}
+
+impl Cache {
+    pub fn new(config: Config) -> Self {
+        let cache = Self {
+            cache: Arc::new(DashMap::new()),
+            expiries: Arc::new(RwLock::new(BTreeMap::new())),
+            config,
+            disk_parity_delta: Arc::new(AtomicU32::new(0)),
+        };
+
+        // try loading from file
+        if let Err(err) = async_block_on(cache.load()) {
+            tracing::warn!("failed to load cache from file: {}", err);
+        };
+
+        cache
+    }
+
+    pub async fn load(&self) -> Result<(), CacheError> {
+        if let Some(filename) = &self.config.filename {
+            let encoded = std::fs::read(filename).context(DiskSnafu)?;
+
+            let cache: HashMap<SystemTime, (Vec<u8>, Vec<u8>)> = bincode_opts()
+                .deserialize(&encoded)
+                .context(DeserializationSnafu)?;
+
+            // inline prune and insert
+            let now = SystemTime::now();
+            for (expiry, (key, value)) in cache {
+                if now < expiry {
+                    self.cache.insert(key.clone(), value);
+                    self.expiries.write().await.insert(expiry, key);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn save(&self) -> Result<(), CacheError> {
+        if let Some(filename) = &self.config.filename {
+            // prune first
+            self.prune().await;
+
+            // serialize
+            let mut cache_to_write = HashMap::new();
+            let expiries = self.expiries.read().await;
+            for (expiry, key) in &*expiries {
+                if let Some(entry) = self.cache.get(key) {
+                    cache_to_write.insert(expiry, (key, entry.value().clone()));
+                } else {
+                    tracing::warn!("key not found in cache: {:?}", key);
+                    Err(CacheError::GeneralCache {
+                        source: Box::new(bincode::ErrorKind::Custom(
+                            "key not found in cache".to_string(),
+                        )),
+                    })?;
+                };
+            }
+
+            let encoded = bincode_opts()
+                .serialize(&cache_to_write)
+                .context(SerializationSnafu)?;
+
+            std::fs::write(filename, encoded).context(DiskSnafu)?;
+        }
+
+        Ok(())
+    }
+
+    async fn prune(&self) {
+        let now = SystemTime::now();
+        let mut expiries = self.expiries.write().await;
+        let mut removed: u32 = 0;
+
+        while let Some((expires, key)) = expiries.pop_first() {
+            if now > expires {
+                self.cache.remove(&key);
+                removed += 1;
+            } else {
+                expiries.insert(expires, key);
+                break;
+            }
+        }
+
+        // todo: test speed on this as opposed to inline
```

Comment:
> as opposed to inline
What would be inlined?

Reply:
I had meant doing a `fetch_add` inside the loop as opposed to tracking how many changes and then doing one atomic add at the end. Definitely wouldn't make a big difference, probably don't have to "todo" this lol, will remove it